### ISSUE TITLE
Cannot get tests discovered

### ISSUE THREAD

**Issue by kentcb** [xunit/xunit#295]
I feel like I must be doing something stupid, but I've followed the docs and still cannot get tests to appear in *Test Explorer* in VS2015 (or 2013). Here's what I did:

* created a new solution with 2 projects: one a PCL and one a normal class library
* added these packages:
  * **xunit** (2.0.0-rc1-build2826)
  * **xunit.abstractions** (2.0.0-rc1-build2826)
  * **xunit.assert** (2.0.0-rc1-build2826)
  * **xunit.core** (2.0.0-rc1-build2826)
  * **xunit.extensibility.core** (2.0.0-rc1-build2826)
  * **xunit.runner.visualstudio** (2.0.0-rc1-build1030)
  * **xunit.runners** (2.0.0-rc1-build2826). I wasn't certain I needed this one, so I only added it when things weren't working without it. The docs seem a little ambiguous as to whether this package is needed when running tests in VS.
* added an empty test
* displayed the test explorer
* rebuilt my solution

The *Output* window for *Tests* just says:
```
------ Discover test started ------
========== Discover test finished: 0 found (0:00:00.145672) ==========
```
Am I doing something wrong here?

**Comment by Clockwork-Muse** [comment]
This appears to be something specific to the rc build.  Going back to the previous build makes them show up again (not a good situation for a release-candidate build to be in, though).

Oh, no, you don't need the xunit.runners if running from visual studio.

**Comment by kentcb** [comment]
Hmm, I just tried rolling my test solution back to beta5, but still no tests are showing...

**Comment by bradwilson** [comment]
I'm at a loss as to what your issue is.

Here is our Visual Studio integration test project: https://github.com/xunit/visualstudio.xunit.integration/

It is linked against the absolute latest bits (nearly complete RC2). This gets run by several people in several different versions of Visual Studio. Personally, I've run these tests in Visual Studio 2013 Professional (with Update 4), Visual Studio 2013 Ultimate (with Update 4), and Visual Studio 2015 Preview (aka CTP 4).

Be aware that Visual Studio 2015 CTP 5 has been reported to be broken by many people with unit testing in general (not just xUnit.net), so don't expect that to work.

**Also, please make sure you've cleaned out Visual Studio's runner cache. If it gets corrupted, Visual Studio will permanently misbehave until it's deleted. To clear the cache, shut down all instances of Visual Studio, then delete the folder `%TEMP%\VisualStudioTestExplorerExtensions` (honestly, it probably wouldn't hurt to delete everything in `%TEMP%` that can be deleted).**

**Comment by kentcb** [comment]
Thanks Brad. Deleting *%TEMP%\VisualStudioTestExplorerExtensions* worked for me in both 2013 and 2015. Do you think it might be worth mentioning this in [this doc](http://xunit.github.io/docs/getting-started.html)?

**Comment by bradwilson** [comment]
I was hoping the problem wouldn't be so widespread, but since you're now like the dozenth person I've told this, I guess my hope was unfounded. :-p

**Comment by bradwilson** [comment]
I've updated this page:

http://xunit.github.io/docs/running-tests-in-vs.html

And I pointed the Getting Started page over here (rather than duplicate the content).

**Comment by kentcb** [comment]
FYI, similar problems using 2.0.0-rc3 with VS2015 CTP6, even after clearing the cache. It just outputs "Discover test started", but never actually discovers any (and seemingly never gives up trying).

**Comment by bradwilson** [comment]
Does it start and never finish discovery (i.e., the "cylon bar" just spins green forever)? If so, that's the problem I had all the time with CTP 5, but haven't seen on the Preview or on CTP 6.

**Comment by kentcb** [comment]
@bradwilson yes, that's exactly what it's doing

**Comment by bradwilson** [comment]
I was never able to figure out what was going on, and the dev team at Microsoft could not repro the issue. All I know is that CTP 4 worked, CTP 5 was broken (built mutiple VMs, all broken), and CTP 6 is working again for me. :(

**Comment by npnelson** [comment]
For what it's worth, I also had all kinds of problems with this and CTP6 (yet when I cloned the VS Integration Test project, those tests discovered/ran fine).  

I finally tracked the problem down to the fact that my own test projects were sitting on a UNC path.  I am running Windows on a Mac via Parallels and by default, new projects were getting dropped into the Parallels shared folder (i.e. \\psf\...)

When I put my project on a local drive, test discovery worked.  So maybe the problem manifests itself when using a UNC path or maybe it is something specific to Parallels?

**Comment by bradwilson** [comment]
If it's really a UNC problem, then you're running afoul of the fact that .NET does not trust code loaded from a UNC path at the same level as code loaded from a local path.

**Comment by kentcb** [comment]
@npnelson yep, you're definitely onto something - thanks. I'm also using parallels and have been struggling on and off with these kinds of issues for a while now. I've just validated your theory and found that the same little test project works (i.e. resolves tests) when on a mapped drive, but not when on a UNC path:

![image](https://cloud.githubusercontent.com/assets/1901832/6423530/6fda4d5e-bf37-11e4-8ff8-860d8d3d775d.png)

Number 2 works even though Y: is just mapped to *\\psf\Home\*.

@bradwilson seems like it would be worth adding another note to the docs about this?

**Comment by lrodrigueza** [comment]
I'm testing the newest xunit release candidates 2.1.0-rc1. These candidates allowed me to resolve a System.Threading.Tasks dependency error that didn't allow my solution to build. However, even after successfully building the solution, the tests are still not discovered. I've tried many times to delete %TEMP%\VisualStudioTestExplorerExtensions file, but every time I re-open VS 2013 this file gets created again. I've also cleared everything out %TEMP% as much as possible. Any ideas?

I have the following packages installed: 
xunit.abstractions v2.0.0
xunit.assert v2.1.0-rc1-build3168
xunit.extensibility.core v2.1.0-rc1-build3168
xunit.runner.console  v2.1.0-rc1-build3168
xunit.runner.visualstudio v2.1.0-rc1-build1124

Output window says
------ Discover test started ------
========== Discover test finished: 0 found (0:00:01.4331433) ==========

**Comment by Banashek** [comment]
+1 to @kentcb 

Running on windows with user home default set to a network path.

Mapping the path, then reopening VS (make sure to re-open the project from the mapped directory!) solved the issue.

**Comment by giobb** [comment]
Hey guys. I haven't had luck with this at all for the EntityFramework code base. I've tried deleting the temp directory as suggested here and one thread somewhere. My codes are not in UNC so I've ran out of any possible culprit.

**Comment by ip28** [comment]
Guys, I am facing this issue in two different machines, both running VS 2015 Enterprise (one has resharper ultimate installed in it, another doesn't).
I am trying to test a controller and the tests are not discoverable. I have tried clearing the temp files, and have restarted Visual Studio too.
It would be great if you can give a workaround or fix for this
![vs2015xunit](https://cloud.githubusercontent.com/assets/8371740/11112716/cbabd864-890d-11e5-81f2-2c62459f55bb.JPG)

**Comment by ip28** [comment]
As mentioned on link http://xunit.github.io/docs/running-tests-in-vs.html
I have installed the xunit runner nguet package for Visual Studio
![xunitrunner](https://cloud.githubusercontent.com/assets/8371740/11112773/66bd91bc-890e-11e5-9457-cde1df8c93c1.JPG)

**Comment by bradwilson** [comment]
@ip28 https://twitter.com/xunit/status/623230451776249856

**Comment by bradwilson** [comment]
@ip28 I've reorganized the doc site to remove that old page, since it wasn't clear that that page was originally intended for people writing desktop .NET tests, not DNX. Now there are only the getting started pages, each labeled with the intended target platform.

![Getting Started](https://cloud.githubusercontent.com/assets/16944/11124577/a113460c-891a-11e5-9ce6-d20164525499.png)

**Comment by turowicz** [comment]
No tests are discovered after following this guide:

https://xunit.github.io/docs/getting-started-desktop.html

Also tried the troubleshooting steps.

**Comment by pantonis** [comment]
same problem here.

**Comment by bradwilson** [comment]
If you're following the desktop instructions for DNX unit tests, you're doing it wrong.

If you're posting about desktop unit tests in a thread about DNX unit tests, you're _also_ doing it wrong.

This is a closed issue. If you have a new problem, start a new issue, and provide complete repro steps. Thanks!

**Comment by turowicz** [comment]
it's not clear that this was a dnx issue

**Comment by oromand** [comment]
Trying to configure xunit on vs 2015 under Paralells; struggled to have my test discovered until I re-opened the solution from the map drive and not from the UNC path.

**Comment by bradwilson** [comment]
@oromand wrote:

> Trying to configure xunit on vs 2015 under Paralells; struggled to have my test discovered until I re-opened the solution from the map drive and not from the UNC path.

Yep, this is a Visual Studio limitation.  xUnit.net does not have any problem opening unit tests from UNCs when using our other runners.

**Comment by kaioamaral** [comment]
I've solved my problem by installing the runners for VS2015 Community.

**Comment by torrequinn** [comment]
I know this is an older post but I had a very similar issue. I had a .NETCore Web project and a .NetCore Library project. I tried t create a Unit test project against each. Following the direction I found (https://docs.microsoft.com/en-us/dotnet/articles/core/testing/unit-testing-with-dotnet-test) I had little issue getting the Library tests to detect and work. The test project that tested the Web app (EXE) however would never detect. I cleared the %Temp% cache as instructed, deleted all BIN and OBJ folders, nothing worked. What solved it for me in the end was this
In the executable project's Json file find the build option that says

"emitEntryPoint": true,

set the value to false, compile the soluton and run the tests, For me the tests then detected. I then set it back to true and from there on it detected new tests appropriately.

flakey!

**Comment by michalsindler** [comment]
Deleting %TEMP%\VisualStudioTestExplorerExtensions worked for me too in VS 2013

**Comment by syclee** [comment]
@torrequinn 's solution worked for me, but only after also deleting the Debug folder entirely and then build. After this I saw the xunit dll's being loaded in the Debug folder (it wasn't before). The tests were discovered after this.

**Comment by blackram** [comment]
What has just worked for me is installing xunit.runner.visualstudio nuget into my test project. Then built.

VS 2017 15.0.0+26228.12

Didn't have the temp folder, tried cleaning etc,  ensuring vs settings (projects and solutions) referencing c:

**Comment by smasherprog** [comment]
Tests not discovered using VS 2015 ...
I really wanted to use XUnit, but it seems very buggy

**Comment by pasisavolainen** [comment]
We have this issue in Vs2015u3. In our situation we have to version control the nugets (no restoration in release build). Nuget added references are accessed through a mapped drive (path w:\references\nuget\xunit.etc) and packages.conf files are not committed. After installing xunit & related and fixing paths in csproj (desktop 4.6 library) to point to our mapped drive locations everything seems to work. Until I remove the 'packages' directory from the solution directory.
After that no test gets discovered.
There are no errors at any point about .props not getting loaded etc.

**Comment by bradwilson** [comment]
@pasisavolainen Visual Studio's runner-via-NuGet requires that the package be installed in the `packages` folder. AFAIK, there is no way to override this behavior.

**Comment by 1TheMuffinMan** [comment]
I tried everything. The only thing that worked was installing `Microsoft.DotNet.InternalAbstractions` from nuget.

**Comment by ReneHerrero** [comment]
This is still happening is VS2017.  I noticed this only happens (in my case) when there are UWP projects in the solution.  Take those out and the tests get discovered again.
Note that I'm also using .NET Standard project and the unit tests are on .NET Core.
Hope it helps.

**Comment by jgmdavies** [comment]
Yep, happening for me in VS2017 Pro.  I can add a new Test project (MSTest) and things work for a while then stop (no tests discovered).  I get the feeling it's after I do something with the Nuget package manager.
I'm sometimes seeing "Discover Test failed " with "given assembly name or codebase was invalid" when I open VS2017.  I've tried most of the suggestions in the forums - very frustrating.

**Comment by drmssst** [comment]
Similar to @blackram adding xunit.runner.visualstudio solved the discoverable issue. I'm running VS2017 EE 15.3.3

**Comment by jgmdavies** [comment]
Thanks Brian – are you going for it??!

 

I think I’m too out-of-date for hardware design these days…  (also Sevenoaks not great commute from here).

**Comment by kiran-bobade** [comment]
I was facing the same issue. Uninstalling MSTest* nuget packages worked for me.
